core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1163,4 +1170,21 @@ public long getLastWriteNanos() {
      * This operation will close the engine if the recovery fails.
      */
     public abstract Engine recoverFromTranslog() throws IOException;
+
+    /**
+     * Called when a refresh includes the location.
+     */
+    public static interface RefreshListener {
+        /**
+         * The location to wait for. Must not return null and should be fast.
+         */
+        Translog.Location location();
```

Comment:
I really wonder if we need to expose the location here at all. I think we can just have a single method / func interface like:

```
void onRefreshed(Location location)
```

and let everything else happen in the listener? Why do we care if we have been forcefully refreshed?